logp/logptest/logger.go

```diff
@@ -0,0 +1,187 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logptest
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.elastic.co/ecszap"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// Logger wraps a *logp.Logger and makes it more suitable for tests.
+// Key features:
+//   - All logs are saved on a single temporary log file
+//   - On failures, the log file is kept and its path printed
+//   - Methods to search and wait for log entries are provided,
+//     they keep track of the offset, ensuring ordering when
+//     when searching for logs
+type Logger struct {
+	*logp.Logger
+	t       *testing.T
```

Comment:
```suggestion
```

I'd prefer if `t` is treated as one of the objects, like `context.Context` that is hierarchical and should not be stored in structs but passed as an argument to functions. This is because it's hierarchical and users of this package may desire to use one logger but multiple `t`s (via `t.Run` subtests).